// @finix-payments/finix/lib/api/devicesApi.js
"use strict";
/**
 * Finix API
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevicesApi = exports.DevicesApiApiKeys = void 0;
const request_1 = __importDefault(require("request"));
const models_1 = require("../model/models");
const models_2 = require("../model/models");
const apis_1 = require("./apis");
let defaultBasePath = 'https://finix.sandbox-payments-api.com';
// ===============================================
// This file is autogenerated - Please do not edit
// ===============================================
var DevicesApiApiKeys;
(function (DevicesApiApiKeys) {
})(DevicesApiApiKeys = exports.DevicesApiApiKeys || (exports.DevicesApiApiKeys = {}));
class DevicesApi {
    constructor(basePathOrUsername, password, basePath) {
        this._basePath = defaultBasePath;
        this._defaultHeaders = {};
        this._useQuerystring = false;
        this.authentications = {
            'default': new models_1.VoidAuth(),
            'BasicAuth': new models_2.HttpBasicAuth(),
        };
        this.interceptors = [];
        if (password) {
            this.username = basePathOrUsername;
            this.password = password;
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    setDefaultAuthentication(auth) {
        this.authentications.default = auth;
    }
    setApiKey(key, value) {
        this.authentications[DevicesApiApiKeys[key]].apiKey = value;
    }
    set username(username) {
        this.authentications.BasicAuth.username = username;
    }
    set password(password) {
        this.authentications.BasicAuth.password = password;
    }
    addInterceptor(interceptor) {
        this.interceptors.push(interceptor);
    }
    /**
     * Helper function.
     * Create a `Device` under a `Merchant`.
     * @summary Create a Device
     * @param merchantId ID of the &#x60;Merchant&#x60; object.
     * @param createDevice
     */
    createHelper(merchantId, createDevice, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/merchants/{merchant_id}/devices'
                .replace('{' + 'merchant_id' + '}', encodeURIComponent(String(merchantId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'merchantId' is not null or undefined
            if (merchantId === null || merchantId === undefined) {
                throw new Error('Required parameter merchantId was null or undefined when calling createMerchantDevice.');
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            localVarHeaderParams['Content-Type'] = "application/hal+json";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            if (createDevice && createDevice.hasOwnProperty('file')) {
                //if (createDevice != undefined && createDevice != null && createDevice.hasOwnProperty('file')){
                localVarRequestOptions.formData = createDevice;
            }
            else {
                localVarRequestOptions.body = models_1.ObjectSerializer.serialize(createDevice, "CreateDevice");
            }
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "Device");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Create a `Device` under a `Merchant`.
     * @summary Create a Device
     * @param merchantId ID of the &#x60;Merchant&#x60; object.
     * @param createDevice
     */
    create(merchantId, createDevice, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.createHelper(merchantId, createDevice, options);
            return responseObject.body;
        });
    }
    /**
     * Create a `Device` under a `Merchant`.
     * @summary Create a Device
     * @param merchantId ID of the &#x60;Merchant&#x60; object.
     * @param createDevice
     */
    createHttp(merchantId, createDevice, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.createHelper(merchantId, createDevice, options);
            return responseObject;
        });
    }
    /**
     * Helper function.
     * Retrieve the details of an existing `Device`.  To check the connectivity of the device, include `?include_connection\\=true\\` at the end of the request endpoint.
     * @summary Fetch a Device

    * @param deviceId ID of the &#x60;Device&#x60;.
    *
    */
    getHelper(deviceId, getDeviceQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/devices/{device_id}'
                .replace('{' + 'device_id' + '}', encodeURIComponent(String(deviceId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'deviceId' is not null or undefined
            if (deviceId === null || deviceId === undefined) {
                throw new Error('Required parameter deviceId was null or undefined when calling getDevice.');
            }
            if (getDeviceQueryParams != undefined) {
                if (getDeviceQueryParams.includeConnection !== undefined) {
                    localVarQueryParameters['include_connection'] = models_1.ObjectSerializer.serialize(getDeviceQueryParams.includeConnection, "boolean");
                }
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "Device");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Retrieve the details of an existing `Device`.  To check the connectivity of the device, include `?include_connection\\=true\\` at the end of the request endpoint.
     * @summary Fetch a Device
     * @param deviceId ID of the &#x60;Device&#x60;.
     *
     */
    get(deviceId, getDeviceQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.getHelper(deviceId, getDeviceQueryParams, options);
            return responseObject.body;
        });
    }
    /**
     * Retrieve the details of an existing `Device`.  To check the connectivity of the device, include `?include_connection\\=true\\` at the end of the request endpoint.
     * @summary Fetch a Device
     * @param deviceId ID of the &#x60;Device&#x60;.
     *
     */
    getHttp(deviceId, getDeviceQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.getHelper(deviceId, getDeviceQueryParams, options);
            return responseObject;
        });
    }
    /**
     * Helper function.
     * Initiate an action on a `Device`. These actions include activation, rebooting, setting an idle message, or deactivate it.
     * @summary Initiate Action on Device
     * @param deviceId ID of the &#x60;Device&#x60;.
     * @param updateDeviceRequest
     */
    updateHelper(deviceId, updateDeviceRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/devices/{device_id}'
                .replace('{' + 'device_id' + '}', encodeURIComponent(String(deviceId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'deviceId' is not null or undefined
            if (deviceId === null || deviceId === undefined) {
                throw new Error('Required parameter deviceId was null or undefined when calling updateDevice.');
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            localVarHeaderParams['Content-Type'] = "application/hal+json";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'PUT',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            if (updateDeviceRequest && updateDeviceRequest.hasOwnProperty('file')) {
                //if (updateDeviceRequest != undefined && updateDeviceRequest != null && updateDeviceRequest.hasOwnProperty('file')){
                localVarRequestOptions.formData = updateDeviceRequest;
            }
            else {
                localVarRequestOptions.body = models_1.ObjectSerializer.serialize(updateDeviceRequest, "UpdateDeviceRequest");
            }
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "Device");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Initiate an action on a `Device`. These actions include activation, rebooting, setting an idle message, or deactivate it.
     * @summary Initiate Action on Device
     * @param deviceId ID of the &#x60;Device&#x60;.
     * @param updateDeviceRequest
     */
    update(deviceId, updateDeviceRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.updateHelper(deviceId, updateDeviceRequest, options);
            return responseObject.body;
        });
    }
    /**
     * Initiate an action on a `Device`. These actions include activation, rebooting, setting an idle message, or deactivate it.
     * @summary Initiate Action on Device
     * @param deviceId ID of the &#x60;Device&#x60;.
     * @param updateDeviceRequest
     */
    updateHttp(deviceId, updateDeviceRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.updateHelper(deviceId, updateDeviceRequest, options);
            return responseObject;
        });
    }
    /**
     * Extracts page and links fields from response body and assigns as properties to finixList
     */
    // @ts-ignore: Not all endpoints have list views
    embeddedHelper(responseObject, dataList) {
        if (responseObject.body.embedded == null || responseObject.body.embedded == undefined) {
            dataList.page = responseObject.body.page;
            dataList.links = responseObject.body.links;
            return dataList;
        }
        const embeddedName = Object.getOwnPropertyNames(responseObject.body.embedded)[0];
        let tempList = responseObject.body.embedded[embeddedName];
        tempList.forEach((item) => { dataList.add(item); });
        dataList.page = responseObject.body.page;
        dataList.links = responseObject.body.links;
        return dataList;
    }
    /**
     * Extracts offset value from response body and determines if end of list has been reached
     */
    // @ts-ignore: Not all endpoints have list views
    getOffsetQueryParam(responseObject, queryParam) {
        queryParam.offset = responseObject.body.page.offset + responseObject.body.page.limit;
        var endReached = false;
        if (responseObject.body.page.offset + responseObject.body.page.limit > responseObject.body.page.count) {
            endReached = true;
        }
        return [queryParam, endReached];
    }
    /**
    * Extracts nextCursor value from response body and determines if end of list has been reached
    */
    // @ts-ignore: Not all endpoints have list views
    getCursorQueryParam(responseObject, queryParam) {
        queryParam.afterCursor = responseObject.body.page.nextCursor;
        var endReached = false;
        if (responseObject.body.page.nextCursor == undefined) {
            endReached = true;
        }
        return [queryParam, endReached];
    }
}
exports.DevicesApi = DevicesApi;
