// @finix-payments/finix/lib/api/settlementsApi.js
"use strict";
/**
 * Finix API
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SettlementsApi = exports.SettlementsApiApiKeys = void 0;
const request_1 = __importDefault(require("request"));
const models_1 = require("../model/models");
const models_2 = require("../model/models");
const apis_1 = require("./apis");
let defaultBasePath = 'https://finix.sandbox-payments-api.com';
// ===============================================
// This file is autogenerated - Please do not edit
// ===============================================
var SettlementsApiApiKeys;
(function (SettlementsApiApiKeys) {
})(SettlementsApiApiKeys = exports.SettlementsApiApiKeys || (exports.SettlementsApiApiKeys = {}));
class SettlementsApi {
    constructor(basePathOrUsername, password, basePath) {
        this._basePath = defaultBasePath;
        this._defaultHeaders = {};
        this._useQuerystring = false;
        this.authentications = {
            'default': new models_1.VoidAuth(),
            'BasicAuth': new models_2.HttpBasicAuth(),
        };
        this.interceptors = [];
        if (password) {
            this.username = basePathOrUsername;
            this.password = password;
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    setDefaultAuthentication(auth) {
        this.authentications.default = auth;
    }
    setApiKey(key, value) {
        this.authentications[SettlementsApiApiKeys[key]].apiKey = value;
    }
    set username(username) {
        this.authentications.BasicAuth.username = username;
    }
    set password(password) {
        this.authentications.BasicAuth.password = password;
    }
    addInterceptor(interceptor) {
        this.interceptors.push(interceptor);
    }
    /**
     * Helper function.
     * Close the currently accruing `settlement`.   Finix, by default, creates accruing `settlements` then closes them based on your payout configurations. Use this endpoint to manually close the currently accruing settlement.  The closed `Settlement` will not accrue any further transactions and gets immediately submitted for approval. - Any refunded `Transfers` get included in `Settlements` as a deduction. - **PENDING** `Transfers` don\'t get included in `Settlements`.  - The `total_amount` minus the `total_fee` equals the `net_amount`. The `net_amount` is the amount in cents that gets deposited into the seller\'s bank account.  Related Guides: [Accruing Settlements](/guides/payouts/accruing-settlements/#closing-an-accruing-settlement)
     * @summary Close Current Active Settlement
     * @param identityId ID of the &#x60;Identity&#x60; for the merchant you want to settle.
     * @param createSettlementRequest
     */
    createHelper(identityId, createSettlementRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/identities/{identity_id}/settlements'
                .replace('{' + 'identity_id' + '}', encodeURIComponent(String(identityId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'identityId' is not null or undefined
            if (identityId === null || identityId === undefined) {
                throw new Error('Required parameter identityId was null or undefined when calling createIdentitySettlement.');
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            localVarHeaderParams['Content-Type'] = "application/hal+json";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            if (createSettlementRequest && createSettlementRequest.hasOwnProperty('file')) {
                //if (createSettlementRequest != undefined && createSettlementRequest != null && createSettlementRequest.hasOwnProperty('file')){
                localVarRequestOptions.formData = createSettlementRequest;
            }
            else {
                localVarRequestOptions.body = models_1.ObjectSerializer.serialize(createSettlementRequest, "CreateSettlementRequest");
            }
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "Settlement");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Close the currently accruing `settlement`.   Finix, by default, creates accruing `settlements` then closes them based on your payout configurations. Use this endpoint to manually close the currently accruing settlement.  The closed `Settlement` will not accrue any further transactions and gets immediately submitted for approval. - Any refunded `Transfers` get included in `Settlements` as a deduction. - **PENDING** `Transfers` don\'t get included in `Settlements`.  - The `total_amount` minus the `total_fee` equals the `net_amount`. The `net_amount` is the amount in cents that gets deposited into the seller\'s bank account.  Related Guides: [Accruing Settlements](/guides/payouts/accruing-settlements/#closing-an-accruing-settlement)
     * @summary Close Current Active Settlement
     * @param identityId ID of the &#x60;Identity&#x60; for the merchant you want to settle.
     * @param createSettlementRequest
     */
    create(identityId, createSettlementRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.createHelper(identityId, createSettlementRequest, options);
            return responseObject.body;
        });
    }
    /**
     * Close the currently accruing `settlement`.   Finix, by default, creates accruing `settlements` then closes them based on your payout configurations. Use this endpoint to manually close the currently accruing settlement.  The closed `Settlement` will not accrue any further transactions and gets immediately submitted for approval. - Any refunded `Transfers` get included in `Settlements` as a deduction. - **PENDING** `Transfers` don\'t get included in `Settlements`.  - The `total_amount` minus the `total_fee` equals the `net_amount`. The `net_amount` is the amount in cents that gets deposited into the seller\'s bank account.  Related Guides: [Accruing Settlements](/guides/payouts/accruing-settlements/#closing-an-accruing-settlement)
     * @summary Close Current Active Settlement
     * @param identityId ID of the &#x60;Identity&#x60; for the merchant you want to settle.
     * @param createSettlementRequest
     */
    createHttp(identityId, createSettlementRequest, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.createHelper(identityId, createSettlementRequest, options);
            return responseObject;
        });
    }
    /**
     * Helper function.
     * Retreive the details of a `Settlement`.
     * @summary Fetch a Settlement
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     */
    getHelper(settlementId, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/settlements/{settlement_id}'
                .replace('{' + 'settlement_id' + '}', encodeURIComponent(String(settlementId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'settlementId' is not null or undefined
            if (settlementId === null || settlementId === undefined) {
                throw new Error('Required parameter settlementId was null or undefined when calling getSettlement.');
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "Settlement");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Retreive the details of a `Settlement`.
     * @summary Fetch a Settlement
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     */
    get(settlementId, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.getHelper(settlementId, options);
            return responseObject.body;
        });
    }
    /**
     * Retreive the details of a `Settlement`.
     * @summary Fetch a Settlement
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     */
    getHttp(settlementId, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.getHelper(settlementId, options);
            return responseObject;
        });
    }
    /**
     * Helper function.
     * List the funding `Transfers` that were created when a `Settlement` was approved that have `type` **CREDIT** or **DEBIT**.
     * @summary List Settlement Funding Transfers

    * @param settlementId ID of &#x60;Settlement&#x60; object.
    *
    */
    listFundingTransfersHelper(settlementId, listSettlementFundingTransfersQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/settlements/{settlement_id}/funding_transfers'
                .replace('{' + 'settlement_id' + '}', encodeURIComponent(String(settlementId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'settlementId' is not null or undefined
            if (settlementId === null || settlementId === undefined) {
                throw new Error('Required parameter settlementId was null or undefined when calling listSettlementFundingTransfers.');
            }
            if (listSettlementFundingTransfersQueryParams != undefined) {
                if (listSettlementFundingTransfersQueryParams.limit !== undefined) {
                    localVarQueryParameters['limit'] = models_1.ObjectSerializer.serialize(listSettlementFundingTransfersQueryParams.limit, "number");
                }
                if (listSettlementFundingTransfersQueryParams.afterCursor !== undefined) {
                    localVarQueryParameters['after_cursor'] = models_1.ObjectSerializer.serialize(listSettlementFundingTransfersQueryParams.afterCursor, "string");
                }
                if (listSettlementFundingTransfersQueryParams.beforeCursor !== undefined) {
                    localVarQueryParameters['before_cursor'] = models_1.ObjectSerializer.serialize(listSettlementFundingTransfersQueryParams.beforeCursor, "string");
                }
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "TransfersList");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * List the funding `Transfers` that were created when a `Settlement` was approved that have `type` **CREDIT** or **DEBIT**.
     * @summary List Settlement Funding Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     *
     */
    listFundingTransfers(settlementId, listSettlementFundingTransfersQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listFundingTransfersHelper(settlementId, listSettlementFundingTransfersQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.listFundingTransfers(settlementId, queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, !reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            return dataList;
        });
    }
    /**
     * List the funding `Transfers` that were created when a `Settlement` was approved that have `type` **CREDIT** or **DEBIT**.
     * @summary List Settlement Funding Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     *
     */
    listFundingTransfersHttp(settlementId, listSettlementFundingTransfersQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listFundingTransfersHelper(settlementId, listSettlementFundingTransfersQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.listFundingTransfers(settlementId, queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            //dataList.hasMore = !reachedEnd;
            return Promise.resolve({ response: responseObject.response, body: dataList, rawBody: responseObject.rawBody });
        });
    }
    /**
     * Helper function.
     * Retrieve a list of every `Transfer` in a `Settlement` that has `type` **DEBIT** or **REFUND**.
     * @summary List Settlement Transfers

    * @param settlementId ID of &#x60;Settlement&#x60; object.
    *
    */
    listTransfersBySettlementIdHelper(settlementId, listSettlementTransfersQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/settlements/{settlement_id}/transfers'
                .replace('{' + 'settlement_id' + '}', encodeURIComponent(String(settlementId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'settlementId' is not null or undefined
            if (settlementId === null || settlementId === undefined) {
                throw new Error('Required parameter settlementId was null or undefined when calling listSettlementTransfers.');
            }
            if (listSettlementTransfersQueryParams != undefined) {
                if (listSettlementTransfersQueryParams.limit !== undefined) {
                    localVarQueryParameters['limit'] = models_1.ObjectSerializer.serialize(listSettlementTransfersQueryParams.limit, "number");
                }
                if (listSettlementTransfersQueryParams.afterCursor !== undefined) {
                    localVarQueryParameters['after_cursor'] = models_1.ObjectSerializer.serialize(listSettlementTransfersQueryParams.afterCursor, "string");
                }
                if (listSettlementTransfersQueryParams.beforeCursor !== undefined) {
                    localVarQueryParameters['before_cursor'] = models_1.ObjectSerializer.serialize(listSettlementTransfersQueryParams.beforeCursor, "string");
                }
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "TransfersList");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Retrieve a list of every `Transfer` in a `Settlement` that has `type` **DEBIT** or **REFUND**.
     * @summary List Settlement Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     *
     */
    listTransfersBySettlementId(settlementId, listSettlementTransfersQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listTransfersBySettlementIdHelper(settlementId, listSettlementTransfersQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.listTransfersBySettlementId(settlementId, queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, !reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            return dataList;
        });
    }
    /**
     * Retrieve a list of every `Transfer` in a `Settlement` that has `type` **DEBIT** or **REFUND**.
     * @summary List Settlement Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     *
     */
    listTransfersBySettlementIdHttp(settlementId, listSettlementTransfersQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listTransfersBySettlementIdHelper(settlementId, listSettlementTransfersQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.listTransfersBySettlementId(settlementId, queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            //dataList.hasMore = !reachedEnd;
            return Promise.resolve({ response: responseObject.response, body: dataList, rawBody: responseObject.rawBody });
        });
    }
    /**
     * Helper function.
     * Retrieve a list of `Settlements`.
     * @summary List All Settlements

    */
    listHelper(listSettlementsQueryParams, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/settlements';
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            if (listSettlementsQueryParams != undefined) {
                if (listSettlementsQueryParams.createdAtGte !== undefined) {
                    localVarQueryParameters['created_at.gte'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.createdAtGte, "string");
                }
                if (listSettlementsQueryParams.createdAtLte !== undefined) {
                    localVarQueryParameters['created_at.lte'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.createdAtLte, "string");
                }
                if (listSettlementsQueryParams.updatedAtGte !== undefined) {
                    localVarQueryParameters['updated_at.gte'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.updatedAtGte, "string");
                }
                if (listSettlementsQueryParams.updatedAtLte !== undefined) {
                    localVarQueryParameters['updated_at.lte'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.updatedAtLte, "string");
                }
                if (listSettlementsQueryParams.id !== undefined) {
                    localVarQueryParameters['id'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.id, "string");
                }
                if (listSettlementsQueryParams.limit !== undefined) {
                    localVarQueryParameters['limit'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.limit, "number");
                }
                if (listSettlementsQueryParams.afterCursor !== undefined) {
                    localVarQueryParameters['after_cursor'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.afterCursor, "string");
                }
                if (listSettlementsQueryParams.beforeCursor !== undefined) {
                    localVarQueryParameters['before_cursor'] = models_1.ObjectSerializer.serialize(listSettlementsQueryParams.beforeCursor, "string");
                }
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                body = models_1.ObjectSerializer.deserialize(body, "SettlementsList");
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Retrieve a list of `Settlements`.
     * @summary List All Settlements
     */
    list(listSettlementsQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listHelper(listSettlementsQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.list(queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, !reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            return dataList;
        });
    }
    /**
     * Retrieve a list of `Settlements`.
     * @summary List All Settlements
     */
    listHttp(listSettlementsQueryParams, options = { headers: {} }) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.listHelper(listSettlementsQueryParams, options);
            // Check if response body has nextCursor property or offset property and extract the corresponding fields
            let reachedEnd;
            const hasNextCursor = (_b = (_a = responseObject.body) === null || _a === void 0 ? void 0 : _a.page) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('nextCursor');
            let queryParam = hasNextCursor ? { afterCursor: '', limit: 20 } : { offset: '', limit: 20 };
            [queryParam, reachedEnd] = hasNextCursor
                ? this.getCursorQueryParam(responseObject, queryParam)
                : this.getOffsetQueryParam(responseObject, queryParam);
            const nextFetch = (limit) => {
                queryParam.limit = limit;
                if (reachedEnd) {
                    throw new RangeError("End of list reached");
                }
                return this.list(queryParam);
            };
            let dataList = new models_1.finixList(nextFetch, reachedEnd);
            dataList = this.embeddedHelper(responseObject, dataList);
            //dataList.hasMore = !reachedEnd;
            return Promise.resolve({ response: responseObject.response, body: dataList, rawBody: responseObject.rawBody });
        });
    }
    /**
     * Helper function.
     * Remove a `Transfer` that makes up a `Settlement`.  As long as the `Settlement` hasn\'t been funded, you can remove the `Transfer` or an array of `Transfers`, along with its corresponding `fee` from the encompassing `Settlement`. - Funding `transfers` can\'t be deleted.   > Per the JSON API for deleting a resource, our API doesn\'t have a response body when removing a `Transfer` from a `Settlement`.
     * @summary Delete Settlement Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     * @param removeSettlementTransfer
     */
    removeTransfersFromSettlementHelper(settlementId, removeSettlementTransfer, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const localVarPath = this.basePath + '/settlements/{settlement_id}/transfers'
                .replace('{' + 'settlement_id' + '}', encodeURIComponent(String(settlementId)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this._defaultHeaders);
            const produces = ['application/hal+json'];
            // give precedence to 'application/json'
            if (produces.indexOf('application/json') >= 0) {
                localVarHeaderParams.Accept = 'application/json';
            }
            else {
                localVarHeaderParams.Accept = produces.join(',');
            }
            let localVarFormParams = {};
            // verify required parameter 'settlementId' is not null or undefined
            if (settlementId === null || settlementId === undefined) {
                throw new Error('Required parameter settlementId was null or undefined when calling removeSettlementTransfers.');
            }
            Object.assign(localVarHeaderParams, options.headers);
            localVarHeaderParams['Finix-Version'] = "2022-02-01";
            localVarHeaderParams['Content-Type'] = "application/hal+json";
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'DELETE',
                qs: localVarQueryParameters,
                headers: localVarHeaderParams,
                uri: localVarPath,
                useQuerystring: this._useQuerystring,
                json: true,
            };
            if (removeSettlementTransfer && removeSettlementTransfer.hasOwnProperty('file')) {
                //if (removeSettlementTransfer != undefined && removeSettlementTransfer != null && removeSettlementTransfer.hasOwnProperty('file')){
                localVarRequestOptions.formData = removeSettlementTransfer;
            }
            else {
                localVarRequestOptions.body = models_1.ObjectSerializer.serialize(removeSettlementTransfer, "RemoveSettlementTransfer");
            }
            let authenticationPromise = Promise.resolve();
            if (this.authentications.BasicAuth.username && this.authentications.BasicAuth.password) {
                authenticationPromise = authenticationPromise.then(() => this.authentications.BasicAuth.applyToRequest(localVarRequestOptions));
            }
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            let interceptorPromise = authenticationPromise;
            for (const interceptor of this.interceptors) {
                interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
            }
            return interceptorPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.formData = localVarFormParams;
                    }
                    else {
                        localVarRequestOptions.form = localVarFormParams;
                    }
                }
                return new Promise((resolve, reject) => {
                    (0, request_1.default)(localVarRequestOptions, (error, response, body) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                                const rawBody = body;
                                resolve({ response: response, body: body, rawBody: rawBody });
                            }
                            else {
                                reject(new apis_1.HttpError(response, body, response.statusCode));
                            }
                        }
                    });
                });
            });
        });
    }
    /**
     * Remove a `Transfer` that makes up a `Settlement`.  As long as the `Settlement` hasn\'t been funded, you can remove the `Transfer` or an array of `Transfers`, along with its corresponding `fee` from the encompassing `Settlement`. - Funding `transfers` can\'t be deleted.   > Per the JSON API for deleting a resource, our API doesn\'t have a response body when removing a `Transfer` from a `Settlement`.
     * @summary Delete Settlement Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     * @param removeSettlementTransfer
     */
    removeTransfersFromSettlement(settlementId, removeSettlementTransfer, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.removeTransfersFromSettlementHelper(settlementId, removeSettlementTransfer, options);
            return responseObject.body;
        });
    }
    /**
     * Remove a `Transfer` that makes up a `Settlement`.  As long as the `Settlement` hasn\'t been funded, you can remove the `Transfer` or an array of `Transfers`, along with its corresponding `fee` from the encompassing `Settlement`. - Funding `transfers` can\'t be deleted.   > Per the JSON API for deleting a resource, our API doesn\'t have a response body when removing a `Transfer` from a `Settlement`.
     * @summary Delete Settlement Transfers
     * @param settlementId ID of &#x60;Settlement&#x60; object.
     * @param removeSettlementTransfer
     */
    removeTransfersFromSettlementHttp(settlementId, removeSettlementTransfer, options = { headers: {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseObject = yield this.removeTransfersFromSettlementHelper(settlementId, removeSettlementTransfer, options);
            return responseObject;
        });
    }
    /**
     * Extracts page and links fields from response body and assigns as properties to finixList
     */
    // @ts-ignore: Not all endpoints have list views
    embeddedHelper(responseObject, dataList) {
        if (responseObject.body.embedded == null || responseObject.body.embedded == undefined) {
            dataList.page = responseObject.body.page;
            dataList.links = responseObject.body.links;
            return dataList;
        }
        const embeddedName = Object.getOwnPropertyNames(responseObject.body.embedded)[0];
        let tempList = responseObject.body.embedded[embeddedName];
        tempList.forEach((item) => { dataList.add(item); });
        dataList.page = responseObject.body.page;
        dataList.links = responseObject.body.links;
        return dataList;
    }
    /**
     * Extracts offset value from response body and determines if end of list has been reached
     */
    // @ts-ignore: Not all endpoints have list views
    getOffsetQueryParam(responseObject, queryParam) {
        queryParam.offset = responseObject.body.page.offset + responseObject.body.page.limit;
        var endReached = false;
        if (responseObject.body.page.offset + responseObject.body.page.limit > responseObject.body.page.count) {
            endReached = true;
        }
        return [queryParam, endReached];
    }
    /**
    * Extracts nextCursor value from response body and determines if end of list has been reached
    */
    // @ts-ignore: Not all endpoints have list views
    getCursorQueryParam(responseObject, queryParam) {
        queryParam.afterCursor = responseObject.body.page.nextCursor;
        var endReached = false;
        if (responseObject.body.page.nextCursor == undefined) {
            endReached = true;
        }
        return [queryParam, endReached];
    }
}
exports.SettlementsApi = SettlementsApi;
