// @finix-payments/finix/lib/test/complianceForms.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("../api");
describe('Compliance Forms API', () => {
    let client;
    let complianceFormId = "cf_fEojUGLjwUiqNTBp68JWq8";
    beforeAll(() => {
        const userName = 'USj46WbwgnjapmdYFnEDP3Ec';
        const password = 'b9b4042c-9621-438d-a84b-8557d4bda84d';
        client = new api_1.Client(userName, password, api_1.Environment.Sandbox);
    });
    test('Test: Fetch a compliance form', () => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b;
        const complianceForm = yield client.ComplianceForms.list(complianceFormId);
        expect(complianceForm.linkedTo).toBe("MUfnskvHiiDgP7x3TVL2LkG3");
        expect(complianceForm.linkedType).toBe("MERCHANT");
        expect(complianceForm.type).toBe("PCI_SAQ_A");
        expect((_a = complianceForm.pciSaqA) === null || _a === void 0 ? void 0 : _a.isAccepted).toBe(false);
        expect(complianceForm.complianceFormTemplate).toBe("cft_wua8ua1yLAcHRK9mx2mF9K");
        expect((_b = complianceForm.files) === null || _b === void 0 ? void 0 : _b.unsignedFile).toBe("FILE_fFGMCY4sxGYTqpjnXh54kC");
    }), 20000);
    test('Test: Update a compliance form', () => __awaiter(void 0, void 0, void 0, function* () {
        var _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        const client2 = new api_1.Client("US8MrUh4Eb1L9Kn6rgSKdem4", "94879501-e840-4263-ae01-b65e10084893");
        const complianceFormId2 = "cf_pzJCPAqrt9Z5653V3iXRDv";
        const updateComplianceFormRequest = {
            pciSaqA: {
                name: "test_node",
                signedAt: "2022-03-18T16:42:55Z",
                userAgent: "Mozilla 5.0(Macintosh; IntelMac OS X 10 _14_6)",
                ipAddress: "42.1.1.113",
                title: "CTO"
            }
        };
        const complianceForm = yield client2.ComplianceForms.update(complianceFormId2, updateComplianceFormRequest);
        expect((_c = complianceForm.pciSaqA) === null || _c === void 0 ? void 0 : _c.name).toBe((_d = updateComplianceFormRequest.pciSaqA) === null || _d === void 0 ? void 0 : _d.name);
        expect((_e = complianceForm.pciSaqA) === null || _e === void 0 ? void 0 : _e.signedAt).toBe((_f = updateComplianceFormRequest.pciSaqA) === null || _f === void 0 ? void 0 : _f.signedAt);
        expect((_g = complianceForm.pciSaqA) === null || _g === void 0 ? void 0 : _g.userAgent).toBe((_h = updateComplianceFormRequest.pciSaqA) === null || _h === void 0 ? void 0 : _h.userAgent);
        expect((_j = complianceForm.pciSaqA) === null || _j === void 0 ? void 0 : _j.title).toBe((_k = updateComplianceFormRequest.pciSaqA) === null || _k === void 0 ? void 0 : _k.title);
        expect((_l = complianceForm.pciSaqA) === null || _l === void 0 ? void 0 : _l.ipAddress).toBe((_m = updateComplianceFormRequest.pciSaqA) === null || _m === void 0 ? void 0 : _m.ipAddress);
        expect(complianceForm.id).toBe(complianceFormId2);
    }), 20000);
});
