// @finix-payments/finix/lib/model/cardPresentInstrumentForm.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CardPresentInstrumentForm = void 0;
/**
* Details the `Payment Instrument` that the `Transfer` debits or credits.
*/
class CardPresentInstrumentForm {
    static getAttributeTypeMap() {
        return CardPresentInstrumentForm.attributeTypeMap;
    }
}
exports.CardPresentInstrumentForm = CardPresentInstrumentForm;
CardPresentInstrumentForm.discriminator = undefined;
CardPresentInstrumentForm.attributeTypeMap = [
    {
        "name": "type",
        "baseName": "type",
        "type": "CardPresentInstrumentForm.TypeEnum"
    },
    {
        "name": "availableAccountType",
        "baseName": "available_account_type",
        "type": "CardPresentInstrumentForm.AvailableAccountTypeEnum"
    },
    {
        "name": "emvData",
        "baseName": "emv_data",
        "type": "string"
    },
    {
        "name": "emvDataKeySerialNumber",
        "baseName": "emv_data_key_serial_number",
        "type": "string"
    },
    {
        "name": "encryptedEmvData",
        "baseName": "encrypted_emv_data",
        "type": "string"
    },
    {
        "name": "encryptedEmvFormat",
        "baseName": "encrypted_emv_format",
        "type": "number"
    },
    {
        "name": "expirationMonth",
        "baseName": "expiration_month",
        "type": "number"
    },
    {
        "name": "expirationYear",
        "baseName": "expiration_year",
        "type": "number"
    },
    {
        "name": "firstName",
        "baseName": "first_name",
        "type": "string"
    },
    {
        "name": "lastName",
        "baseName": "last_name",
        "type": "string"
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "Name"
    },
    {
        "name": "number",
        "baseName": "number",
        "type": "string"
    }
];
(function (CardPresentInstrumentForm) {
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["BankAccount"] = 'BANK_ACCOUNT'] = "BankAccount";
        TypeEnum[TypeEnum["Virtual"] = 'VIRTUAL'] = "Virtual";
        TypeEnum[TypeEnum["Token"] = 'TOKEN'] = "Token";
        TypeEnum[TypeEnum["SwipedPaymentCard"] = 'SWIPED_PAYMENT_CARD'] = "SwipedPaymentCard";
        TypeEnum[TypeEnum["PaymentCardPresent"] = 'PAYMENT_CARD_PRESENT'] = "PaymentCardPresent";
        TypeEnum[TypeEnum["PaymentCard"] = 'PAYMENT_CARD'] = "PaymentCard";
        TypeEnum[TypeEnum["VantivOmniToken"] = 'VANTIV_OMNI_TOKEN'] = "VantivOmniToken";
    })(TypeEnum = CardPresentInstrumentForm.TypeEnum || (CardPresentInstrumentForm.TypeEnum = {}));
    let AvailableAccountTypeEnum;
    (function (AvailableAccountTypeEnum) {
        AvailableAccountTypeEnum[AvailableAccountTypeEnum["Balance"] = 'BALANCE'] = "Balance";
        AvailableAccountTypeEnum[AvailableAccountTypeEnum["Ledgered"] = 'LEDGERED'] = "Ledgered";
    })(AvailableAccountTypeEnum = CardPresentInstrumentForm.AvailableAccountTypeEnum || (CardPresentInstrumentForm.AvailableAccountTypeEnum = {}));
})(CardPresentInstrumentForm = exports.CardPresentInstrumentForm || (exports.CardPresentInstrumentForm = {}));
