// @finix-payments/finix/lib/model/complianceForm.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComplianceForm = void 0;
class ComplianceForm {
    static getAttributeTypeMap() {
        return ComplianceForm.attributeTypeMap;
    }
}
exports.ComplianceForm = ComplianceForm;
ComplianceForm.discriminator = undefined;
ComplianceForm.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "complianceFormTemplate",
        "baseName": "compliance_form_template",
        "type": "string"
    },
    {
        "name": "dueAt",
        "baseName": "due_at",
        "type": "Date"
    },
    {
        "name": "files",
        "baseName": "files",
        "type": "ComplianceFormFiles"
    },
    {
        "name": "linkedTo",
        "baseName": "linked_to",
        "type": "string"
    },
    {
        "name": "linkedType",
        "baseName": "linked_type",
        "type": "string"
    },
    {
        "name": "pciSaqA",
        "baseName": "pci_saq_a",
        "type": "ComplianceFormPciSaqA"
    },
    {
        "name": "state",
        "baseName": "state",
        "type": "ComplianceForm.StateEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "ComplianceForm.TypeEnum"
    },
    {
        "name": "validFrom",
        "baseName": "valid_from",
        "type": "Date"
    },
    {
        "name": "validUntil",
        "baseName": "valid_until",
        "type": "string"
    },
    {
        "name": "version",
        "baseName": "version",
        "type": "string"
    }
];
(function (ComplianceForm) {
    let StateEnum;
    (function (StateEnum) {
        StateEnum[StateEnum["Pending"] = 'PENDING'] = "Pending";
        StateEnum[StateEnum["Completed"] = 'COMPLETED'] = "Completed";
        StateEnum[StateEnum["Invalid"] = 'INVALID'] = "Invalid";
        StateEnum[StateEnum["Incomplete"] = 'INCOMPLETE'] = "Incomplete";
    })(StateEnum = ComplianceForm.StateEnum || (ComplianceForm.StateEnum = {}));
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["PciSaqA"] = 'PCI_SAQ_A'] = "PciSaqA";
    })(TypeEnum = ComplianceForm.TypeEnum || (ComplianceForm.TypeEnum = {}));
})(ComplianceForm = exports.ComplianceForm || (exports.ComplianceForm = {}));
