// @finix-payments/finix/lib/model/createPaymentInstrumentRequest.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreatePaymentInstrumentRequest = void 0;
class CreatePaymentInstrumentRequest {
    constructor() {
        /**
        * Set to **true** if you want to request a bank account validation. Default value is **false**.
        */
        this['attemptBankAccountValidationCheck'] = false;
    }
    static getAttributeTypeMap() {
        return CreatePaymentInstrumentRequest.attributeTypeMap;
    }
}
exports.CreatePaymentInstrumentRequest = CreatePaymentInstrumentRequest;
CreatePaymentInstrumentRequest.discriminator = undefined;
CreatePaymentInstrumentRequest.attributeTypeMap = [
    {
        "name": "address",
        "baseName": "address",
        "type": "CreatePaymentInstrumentRequestAddress"
    },
    {
        "name": "expirationMonth",
        "baseName": "expiration_month",
        "type": "number"
    },
    {
        "name": "expirationYear",
        "baseName": "expiration_year",
        "type": "number"
    },
    {
        "name": "identity",
        "baseName": "identity",
        "type": "string"
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "string"
    },
    {
        "name": "number",
        "baseName": "number",
        "type": "string"
    },
    {
        "name": "securityCode",
        "baseName": "security_code",
        "type": "string"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "CreatePaymentInstrumentRequest.TypeEnum"
    },
    {
        "name": "thirdPartyToken",
        "baseName": "third_party_token",
        "type": "string"
    },
    {
        "name": "accountNumber",
        "baseName": "account_number",
        "type": "string"
    },
    {
        "name": "accountType",
        "baseName": "account_type",
        "type": "CreatePaymentInstrumentRequest.AccountTypeEnum"
    },
    {
        "name": "attemptBankAccountValidationCheck",
        "baseName": "attempt_bank_account_validation_check",
        "type": "boolean"
    },
    {
        "name": "bankCode",
        "baseName": "bank_code",
        "type": "string"
    },
    {
        "name": "country",
        "baseName": "country",
        "type": "string"
    },
    {
        "name": "token",
        "baseName": "token",
        "type": "string"
    },
    {
        "name": "merchantIdentity",
        "baseName": "merchant_identity",
        "type": "string"
    }
];
(function (CreatePaymentInstrumentRequest) {
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["PaymentCard"] = 'PAYMENT_CARD'] = "PaymentCard";
        TypeEnum[TypeEnum["ApplePay"] = 'APPLE_PAY'] = "ApplePay";
        TypeEnum[TypeEnum["GooglePay"] = 'GOOGLE_PAY'] = "GooglePay";
        TypeEnum[TypeEnum["Token"] = 'TOKEN'] = "Token";
        TypeEnum[TypeEnum["BankAccount"] = 'BANK_ACCOUNT'] = "BankAccount";
    })(TypeEnum = CreatePaymentInstrumentRequest.TypeEnum || (CreatePaymentInstrumentRequest.TypeEnum = {}));
    let AccountTypeEnum;
    (function (AccountTypeEnum) {
        AccountTypeEnum[AccountTypeEnum["Checking"] = 'CHECKING'] = "Checking";
        AccountTypeEnum[AccountTypeEnum["Savings"] = 'SAVINGS'] = "Savings";
    })(AccountTypeEnum = CreatePaymentInstrumentRequest.AccountTypeEnum || (CreatePaymentInstrumentRequest.AccountTypeEnum = {}));
})(CreatePaymentInstrumentRequest = exports.CreatePaymentInstrumentRequest || (exports.CreatePaymentInstrumentRequest = {}));
