// @finix-payments/finix/lib/model/createFeeRequest.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateFeeRequest = void 0;
class CreateFeeRequest {
    static getAttributeTypeMap() {
        return CreateFeeRequest.attributeTypeMap;
    }
}
exports.CreateFeeRequest = CreateFeeRequest;
CreateFeeRequest.discriminator = undefined;
CreateFeeRequest.attributeTypeMap = [
    {
        "name": "amount",
        "baseName": "amount",
        "type": "number"
    },
    {
        "name": "currency",
        "baseName": "currency",
        "type": "Currency"
    },
    {
        "name": "feeSubtype",
        "baseName": "fee_subtype",
        "type": "CreateFeeRequest.FeeSubtypeEnum"
    },
    {
        "name": "feeType",
        "baseName": "fee_type",
        "type": "CreateFeeRequest.FeeTypeEnum"
    },
    {
        "name": "label",
        "baseName": "label",
        "type": "string"
    },
    {
        "name": "linkedId",
        "baseName": "linked_id",
        "type": "string"
    },
    {
        "name": "linkedType",
        "baseName": "linked_type",
        "type": "CreateFeeRequest.LinkedTypeEnum"
    },
    {
        "name": "merchantId",
        "baseName": "merchant_id",
        "type": "string"
    },
    {
        "name": "settlementDelayDays",
        "baseName": "settlement_delay_days",
        "type": "number"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    }
];
(function (CreateFeeRequest) {
    let FeeSubtypeEnum;
    (function (FeeSubtypeEnum) {
        FeeSubtypeEnum[FeeSubtypeEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FeeSubtypeEnum = CreateFeeRequest.FeeSubtypeEnum || (CreateFeeRequest.FeeSubtypeEnum = {}));
    let FeeTypeEnum;
    (function (FeeTypeEnum) {
        FeeTypeEnum[FeeTypeEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FeeTypeEnum = CreateFeeRequest.FeeTypeEnum || (CreateFeeRequest.FeeTypeEnum = {}));
    let LinkedTypeEnum;
    (function (LinkedTypeEnum) {
        LinkedTypeEnum[LinkedTypeEnum["Application"] = 'APPLICATION'] = "Application";
        LinkedTypeEnum[LinkedTypeEnum["Platform"] = 'PLATFORM'] = "Platform";
        LinkedTypeEnum[LinkedTypeEnum["Subscription"] = 'SUBSCRIPTION'] = "Subscription";
        LinkedTypeEnum[LinkedTypeEnum["Transfer"] = 'TRANSFER'] = "Transfer";
    })(LinkedTypeEnum = CreateFeeRequest.LinkedTypeEnum || (CreateFeeRequest.LinkedTypeEnum = {}));
})(CreateFeeRequest = exports.CreateFeeRequest || (exports.CreateFeeRequest = {}));
