// @finix-payments/finix/lib/model/createFileRequest.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateFileRequest = void 0;
class CreateFileRequest {
    static getAttributeTypeMap() {
        return CreateFileRequest.attributeTypeMap;
    }
}
exports.CreateFileRequest = CreateFileRequest;
CreateFileRequest.discriminator = undefined;
CreateFileRequest.attributeTypeMap = [
    {
        "name": "displayName",
        "baseName": "display_name",
        "type": "string"
    },
    {
        "name": "linkedTo",
        "baseName": "linked_to",
        "type": "string"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "CreateFileRequest.TypeEnum"
    }
];
(function (CreateFileRequest) {
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["DriversLicenseFront"] = 'DRIVERS_LICENSE_FRONT'] = "DriversLicenseFront";
        TypeEnum[TypeEnum["DriversLicenseBack"] = 'DRIVERS_LICENSE_BACK'] = "DriversLicenseBack";
        TypeEnum[TypeEnum["IdentificationCardFront"] = 'IDENTIFICATION_CARD_FRONT'] = "IdentificationCardFront";
        TypeEnum[TypeEnum["IdentificationCardBack"] = 'IDENTIFICATION_CARD_BACK'] = "IdentificationCardBack";
        TypeEnum[TypeEnum["BankStatement"] = 'BANK_STATEMENT'] = "BankStatement";
        TypeEnum[TypeEnum["TaxDocument"] = 'TAX_DOCUMENT'] = "TaxDocument";
        TypeEnum[TypeEnum["BusinessRegistration"] = 'BUSINESS_REGISTRATION'] = "BusinessRegistration";
        TypeEnum[TypeEnum["BusinessAddressVerification"] = 'BUSINESS_ADDRESS_VERIFICATION'] = "BusinessAddressVerification";
        TypeEnum[TypeEnum["Other"] = 'OTHER'] = "Other";
        TypeEnum[TypeEnum["PciDocument"] = 'PCI_DOCUMENT'] = "PciDocument";
        TypeEnum[TypeEnum["Passport"] = 'PASSPORT'] = "Passport";
    })(TypeEnum = CreateFileRequest.TypeEnum || (CreateFileRequest.TypeEnum = {}));
})(CreateFileRequest = exports.CreateFileRequest || (exports.CreateFileRequest = {}));
