// @finix-payments/finix/lib/model/createDevice.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateDevice = void 0;
/**
*
*/
class CreateDevice {
    static getAttributeTypeMap() {
        return CreateDevice.attributeTypeMap;
    }
}
exports.CreateDevice = CreateDevice;
CreateDevice.discriminator = undefined;
CreateDevice.attributeTypeMap = [
    {
        "name": "configuration",
        "baseName": "configuration",
        "type": "ConfigurationDetails"
    },
    {
        "name": "description",
        "baseName": "description",
        "type": "string"
    },
    {
        "name": "model",
        "baseName": "model",
        "type": "CreateDevice.ModelEnum"
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "string"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    }
];
(function (CreateDevice) {
    let ModelEnum;
    (function (ModelEnum) {
        ModelEnum[ModelEnum["Bbpos"] = 'BBPOS'] = "Bbpos";
        ModelEnum[ModelEnum["Mx915"] = 'MX915'] = "Mx915";
        ModelEnum[ModelEnum["Mx925"] = 'MX925'] = "Mx925";
        ModelEnum[ModelEnum["Ipp320"] = 'IPP320'] = "Ipp320";
        ModelEnum[ModelEnum["Ipp350"] = 'IPP350'] = "Ipp350";
        ModelEnum[ModelEnum["Isc250"] = 'ISC250'] = "Isc250";
        ModelEnum[ModelEnum["Isc480"] = 'ISC480'] = "Isc480";
        ModelEnum[ModelEnum["Ismp4"] = 'ISMP4'] = "Ismp4";
        ModelEnum[ModelEnum["Android"] = 'ANDROID'] = "Android";
        ModelEnum[ModelEnum["Lane3000"] = 'LANE_3000'] = "Lane3000";
        ModelEnum[ModelEnum["Link2500"] = 'LINK_2500'] = "Link2500";
    })(ModelEnum = CreateDevice.ModelEnum || (CreateDevice.ModelEnum = {}));
})(CreateDevice = exports.CreateDevice || (exports.CreateDevice = {}));
