// @finix-payments/finix/lib/model/disputeEvidence.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisputeEvidence = void 0;
class DisputeEvidence {
    static getAttributeTypeMap() {
        return DisputeEvidence.attributeTypeMap;
    }
}
exports.DisputeEvidence = DisputeEvidence;
DisputeEvidence.discriminator = undefined;
DisputeEvidence.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "dispute",
        "baseName": "dispute",
        "type": "string"
    },
    {
        "name": "state",
        "baseName": "state",
        "type": "DisputeEvidence.StateEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "links",
        "baseName": "_links",
        "type": "DisputeEvidenceLinks"
    }
];
(function (DisputeEvidence) {
    let StateEnum;
    (function (StateEnum) {
        StateEnum[StateEnum["Pending"] = 'PENDING'] = "Pending";
        StateEnum[StateEnum["Succeeded"] = 'SUCCEEDED'] = "Succeeded";
        StateEnum[StateEnum["Canceled"] = 'CANCELED'] = "Canceled";
        StateEnum[StateEnum["Failed"] = 'FAILED'] = "Failed";
    })(StateEnum = DisputeEvidence.StateEnum || (DisputeEvidence.StateEnum = {}));
})(DisputeEvidence = exports.DisputeEvidence || (exports.DisputeEvidence = {}));
