// @finix-payments/finix/lib/model/dispute.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dispute = void 0;
/**
* A `Dispute` objected created for a chargeback or customer disputes.
*/
class Dispute {
    static getAttributeTypeMap() {
        return Dispute.attributeTypeMap;
    }
}
exports.Dispute = Dispute;
Dispute.discriminator = undefined;
Dispute.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "action",
        "baseName": "action",
        "type": "string"
    },
    {
        "name": "amount",
        "baseName": "amount",
        "type": "number"
    },
    {
        "name": "application",
        "baseName": "application",
        "type": "string"
    },
    {
        "name": "disputeDetails",
        "baseName": "dispute_details",
        "type": "{ [key: string]: any; }"
    },
    {
        "name": "identity",
        "baseName": "identity",
        "type": "string"
    },
    {
        "name": "message",
        "baseName": "message",
        "type": "string"
    },
    {
        "name": "occurredAt",
        "baseName": "occurred_at",
        "type": "Date"
    },
    {
        "name": "reason",
        "baseName": "reason",
        "type": "Dispute.ReasonEnum"
    },
    {
        "name": "respondBy",
        "baseName": "respond_by",
        "type": "Date"
    },
    {
        "name": "state",
        "baseName": "state",
        "type": "Dispute.StateEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "transfer",
        "baseName": "transfer",
        "type": "string"
    },
    {
        "name": "links",
        "baseName": "_links",
        "type": "DisputeLinks"
    }
];
(function (Dispute) {
    let ReasonEnum;
    (function (ReasonEnum) {
        ReasonEnum[ReasonEnum["Clerical"] = 'CLERICAL'] = "Clerical";
        ReasonEnum[ReasonEnum["Fraud"] = 'FRAUD'] = "Fraud";
        ReasonEnum[ReasonEnum["Inquiry"] = 'INQUIRY'] = "Inquiry";
        ReasonEnum[ReasonEnum["Quality"] = 'QUALITY'] = "Quality";
        ReasonEnum[ReasonEnum["Technical"] = 'TECHNICAL'] = "Technical";
    })(ReasonEnum = Dispute.ReasonEnum || (Dispute.ReasonEnum = {}));
    let StateEnum;
    (function (StateEnum) {
        StateEnum[StateEnum["Inquiry"] = 'INQUIRY'] = "Inquiry";
        StateEnum[StateEnum["Pending"] = 'PENDING'] = "Pending";
        StateEnum[StateEnum["Won"] = 'WON'] = "Won";
        StateEnum[StateEnum["Lost"] = 'LOST'] = "Lost";
        StateEnum[StateEnum["Arbitration"] = 'ARBITRATION'] = "Arbitration";
    })(StateEnum = Dispute.StateEnum || (Dispute.StateEnum = {}));
})(Dispute = exports.Dispute || (exports.Dispute = {}));
