// @finix-payments/finix/lib/model/operationKey.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationKey = void 0;
/**
* Details the operation that\'ll be performed in the transaction (Card present transactions only) .
*/
var OperationKey;
(function (OperationKey) {
    OperationKey[OperationKey["PushToCard"] = 'PUSH_TO_CARD'] = "PushToCard";
    OperationKey[OperationKey["PullFromCard"] = 'PULL_FROM_CARD'] = "PullFromCard";
    OperationKey[OperationKey["CardPresentDebit"] = 'CARD_PRESENT_DEBIT'] = "CardPresentDebit";
    OperationKey[OperationKey["CardPresentUnreferencedRefund"] = 'CARD_PRESENT_UNREFERENCED_REFUND'] = "CardPresentUnreferencedRefund";
    OperationKey[OperationKey["Sale"] = 'SALE'] = "Sale";
    OperationKey[OperationKey["UnreferencedRefund"] = 'UNREFERENCED_REFUND'] = "UnreferencedRefund";
    OperationKey[OperationKey["MerchantCreditAdjustment"] = 'MERCHANT_CREDIT_ADJUSTMENT'] = "MerchantCreditAdjustment";
    OperationKey[OperationKey["MerchantDebitAdjustment"] = 'MERCHANT_DEBIT_ADJUSTMENT'] = "MerchantDebitAdjustment";
    OperationKey[OperationKey["CardPresentAuthorization"] = 'CARD_PRESENT_AUTHORIZATION'] = "CardPresentAuthorization";
})(OperationKey = exports.OperationKey || (exports.OperationKey = {}));
