// @finix-payments/finix/lib/model/paymentInstrument.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentInstrument = void 0;
/**
*
*/
class PaymentInstrument {
    static getAttributeTypeMap() {
        return PaymentInstrument.attributeTypeMap;
    }
}
exports.PaymentInstrument = PaymentInstrument;
PaymentInstrument.discriminator = undefined;
PaymentInstrument.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "address",
        "baseName": "address",
        "type": "Address"
    },
    {
        "name": "addressVerification",
        "baseName": "address_verification",
        "type": "PaymentInstrument.AddressVerificationEnum"
    },
    {
        "name": "application",
        "baseName": "application",
        "type": "string"
    },
    {
        "name": "bin",
        "baseName": "bin",
        "type": "string"
    },
    {
        "name": "brand",
        "baseName": "brand",
        "type": "PaymentInstrument.BrandEnum"
    },
    {
        "name": "cardName",
        "baseName": "card_name",
        "type": "string"
    },
    {
        "name": "cardType",
        "baseName": "card_type",
        "type": "PaymentInstrument.CardTypeEnum"
    },
    {
        "name": "currency",
        "baseName": "currency",
        "type": "Currency"
    },
    {
        "name": "enabled",
        "baseName": "enabled",
        "type": "boolean"
    },
    {
        "name": "expirationMonth",
        "baseName": "expiration_month",
        "type": "number"
    },
    {
        "name": "expirationYear",
        "baseName": "expiration_year",
        "type": "number"
    },
    {
        "name": "fastFundsIndicator",
        "baseName": "fast_funds_indicator",
        "type": "string"
    },
    {
        "name": "fingerprint",
        "baseName": "fingerprint",
        "type": "string"
    },
    {
        "name": "identity",
        "baseName": "identity",
        "type": "string"
    },
    {
        "name": "instrumentType",
        "baseName": "instrument_type",
        "type": "PaymentInstrument.InstrumentTypeEnum"
    },
    {
        "name": "issuerCountry",
        "baseName": "issuer_country",
        "type": "PaymentInstrument.IssuerCountryEnum"
    },
    {
        "name": "lastFour",
        "baseName": "last_four",
        "type": "string"
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "string"
    },
    {
        "name": "onlineGambingBlockIndicator",
        "baseName": "online_gambing_block_indicator",
        "type": "string"
    },
    {
        "name": "payloadType",
        "baseName": "payload_type",
        "type": "PaymentInstrument.PayloadTypeEnum"
    },
    {
        "name": "pushFundsBlockIndicator",
        "baseName": "push_funds_block_indicator",
        "type": "string"
    },
    {
        "name": "securityCodeVerification",
        "baseName": "security_code_verification",
        "type": "PaymentInstrument.SecurityCodeVerificationEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "PaymentInstrument.TypeEnum"
    },
    {
        "name": "links",
        "baseName": "_links",
        "type": "PaymentInstrumentLinks"
    },
    {
        "name": "accountType",
        "baseName": "account_type",
        "type": "PaymentInstrument.AccountTypeEnum"
    },
    {
        "name": "bankAccountValidationCheck",
        "baseName": "bank_account_validation_check",
        "type": "PaymentInstrument.BankAccountValidationCheckEnum"
    },
    {
        "name": "bankCode",
        "baseName": "bank_code",
        "type": "string"
    },
    {
        "name": "country",
        "baseName": "country",
        "type": "Country"
    },
    {
        "name": "maskedAccountNumber",
        "baseName": "masked_account_number",
        "type": "string"
    }
];
(function (PaymentInstrument) {
    let AddressVerificationEnum;
    (function (AddressVerificationEnum) {
        AddressVerificationEnum[AddressVerificationEnum["PostalCodeAndStreetMatch"] = 'POSTAL_CODE_AND_STREET_MATCH'] = "PostalCodeAndStreetMatch";
        AddressVerificationEnum[AddressVerificationEnum["StreetMatch"] = 'STREET_MATCH'] = "StreetMatch";
        AddressVerificationEnum[AddressVerificationEnum["NoAddress"] = 'NO_ADDRESS'] = "NoAddress";
        AddressVerificationEnum[AddressVerificationEnum["PostalCodeMatch"] = 'POSTAL_CODE_MATCH'] = "PostalCodeMatch";
        AddressVerificationEnum[AddressVerificationEnum["NoMatch"] = 'NO_MATCH'] = "NoMatch";
        AddressVerificationEnum[AddressVerificationEnum["Unknown"] = 'UNKNOWN'] = "Unknown";
        AddressVerificationEnum[AddressVerificationEnum["NotSupported"] = 'NOT_SUPPORTED'] = "NotSupported";
    })(AddressVerificationEnum = PaymentInstrument.AddressVerificationEnum || (PaymentInstrument.AddressVerificationEnum = {}));
    let BrandEnum;
    (function (BrandEnum) {
        BrandEnum[BrandEnum["Uatp"] = 'UATP'] = "Uatp";
        BrandEnum[BrandEnum["ChinaTUnion"] = 'CHINA_T_UNION'] = "ChinaTUnion";
        BrandEnum[BrandEnum["Lankapay"] = 'LANKAPAY'] = "Lankapay";
        BrandEnum[BrandEnum["Mir"] = 'MIR'] = "Mir";
        BrandEnum[BrandEnum["Mastercard"] = 'MASTERCARD'] = "Mastercard";
        BrandEnum[BrandEnum["Dankort"] = 'DANKORT'] = "Dankort";
        BrandEnum[BrandEnum["Interpayment"] = 'INTERPAYMENT'] = "Interpayment";
        BrandEnum[BrandEnum["Unknown"] = 'UNKNOWN'] = "Unknown";
        BrandEnum[BrandEnum["Verve"] = 'VERVE'] = "Verve";
        BrandEnum[BrandEnum["Visa"] = 'VISA'] = "Visa";
        BrandEnum[BrandEnum["Troy"] = 'TROY'] = "Troy";
        BrandEnum[BrandEnum["Discover"] = 'DISCOVER'] = "Discover";
        BrandEnum[BrandEnum["AmericanExpress"] = 'AMERICAN_EXPRESS'] = "AmericanExpress";
        BrandEnum[BrandEnum["Instapayment"] = 'INSTAPAYMENT'] = "Instapayment";
        BrandEnum[BrandEnum["DinersClubInternational"] = 'DINERS_CLUB_INTERNATIONAL'] = "DinersClubInternational";
        BrandEnum[BrandEnum["Maestro"] = 'MAESTRO'] = "Maestro";
        BrandEnum[BrandEnum["DinersClub"] = 'DINERS_CLUB'] = "DinersClub";
        BrandEnum[BrandEnum["ChinaUnionPay"] = 'CHINA_UNION_PAY'] = "ChinaUnionPay";
        BrandEnum[BrandEnum["Rupay"] = 'RUPAY'] = "Rupay";
        BrandEnum[BrandEnum["Jcb"] = 'JCB'] = "Jcb";
    })(BrandEnum = PaymentInstrument.BrandEnum || (PaymentInstrument.BrandEnum = {}));
    let CardTypeEnum;
    (function (CardTypeEnum) {
        CardTypeEnum[CardTypeEnum["Credit"] = 'CREDIT'] = "Credit";
        CardTypeEnum[CardTypeEnum["HsaFsa"] = 'HSA_FSA'] = "HsaFsa";
        CardTypeEnum[CardTypeEnum["NonReloadablePrepaid"] = 'NON_RELOADABLE_PREPAID'] = "NonReloadablePrepaid";
        CardTypeEnum[CardTypeEnum["Debit"] = 'DEBIT'] = "Debit";
        CardTypeEnum[CardTypeEnum["Unknown"] = 'UNKNOWN'] = "Unknown";
        CardTypeEnum[CardTypeEnum["ReloadablePrepaid"] = 'RELOADABLE_PREPAID'] = "ReloadablePrepaid";
    })(CardTypeEnum = PaymentInstrument.CardTypeEnum || (PaymentInstrument.CardTypeEnum = {}));
    let InstrumentTypeEnum;
    (function (InstrumentTypeEnum) {
        InstrumentTypeEnum[InstrumentTypeEnum["VantivOmniToken"] = 'VANTIV_OMNI_TOKEN'] = "VantivOmniToken";
        InstrumentTypeEnum[InstrumentTypeEnum["PaymentCard"] = 'PAYMENT_CARD'] = "PaymentCard";
        InstrumentTypeEnum[InstrumentTypeEnum["ApplePay"] = 'APPLE_PAY'] = "ApplePay";
        InstrumentTypeEnum[InstrumentTypeEnum["GooglePay"] = 'GOOGLE_PAY'] = "GooglePay";
        InstrumentTypeEnum[InstrumentTypeEnum["Token"] = 'TOKEN'] = "Token";
        InstrumentTypeEnum[InstrumentTypeEnum["Virtual"] = 'VIRTUAL'] = "Virtual";
        InstrumentTypeEnum[InstrumentTypeEnum["PaymentCardPresent"] = 'PAYMENT_CARD_PRESENT'] = "PaymentCardPresent";
        InstrumentTypeEnum[InstrumentTypeEnum["BankAccount"] = 'BANK_ACCOUNT'] = "BankAccount";
        InstrumentTypeEnum[InstrumentTypeEnum["SwipedPaymentCard"] = 'SWIPED_PAYMENT_CARD'] = "SwipedPaymentCard";
    })(InstrumentTypeEnum = PaymentInstrument.InstrumentTypeEnum || (PaymentInstrument.InstrumentTypeEnum = {}));
    let IssuerCountryEnum;
    (function (IssuerCountryEnum) {
        IssuerCountryEnum[IssuerCountryEnum["Usa"] = 'USA'] = "Usa";
        IssuerCountryEnum[IssuerCountryEnum["Unknown"] = 'UNKNOWN'] = "Unknown";
        IssuerCountryEnum[IssuerCountryEnum["NonUsa"] = 'NON_USA'] = "NonUsa";
    })(IssuerCountryEnum = PaymentInstrument.IssuerCountryEnum || (PaymentInstrument.IssuerCountryEnum = {}));
    let PayloadTypeEnum;
    (function (PayloadTypeEnum) {
        PayloadTypeEnum[PayloadTypeEnum["Destination"] = 'DESTINATION'] = "Destination";
        PayloadTypeEnum[PayloadTypeEnum["Source"] = 'SOURCE'] = "Source";
    })(PayloadTypeEnum = PaymentInstrument.PayloadTypeEnum || (PaymentInstrument.PayloadTypeEnum = {}));
    let SecurityCodeVerificationEnum;
    (function (SecurityCodeVerificationEnum) {
        SecurityCodeVerificationEnum[SecurityCodeVerificationEnum["Unmatched"] = 'UNMATCHED'] = "Unmatched";
        SecurityCodeVerificationEnum[SecurityCodeVerificationEnum["Matched"] = 'MATCHED'] = "Matched";
        SecurityCodeVerificationEnum[SecurityCodeVerificationEnum["Unknown"] = 'UNKNOWN'] = "Unknown";
    })(SecurityCodeVerificationEnum = PaymentInstrument.SecurityCodeVerificationEnum || (PaymentInstrument.SecurityCodeVerificationEnum = {}));
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["VantivOmniToken"] = 'VANTIV_OMNI_TOKEN'] = "VantivOmniToken";
        TypeEnum[TypeEnum["PaymentCard"] = 'PAYMENT_CARD'] = "PaymentCard";
        TypeEnum[TypeEnum["ApplePay"] = 'APPLE_PAY'] = "ApplePay";
        TypeEnum[TypeEnum["GooglePay"] = 'GOOGLE_PAY'] = "GooglePay";
        TypeEnum[TypeEnum["Token"] = 'TOKEN'] = "Token";
        TypeEnum[TypeEnum["Virtual"] = 'VIRTUAL'] = "Virtual";
        TypeEnum[TypeEnum["PaymentCardPresent"] = 'PAYMENT_CARD_PRESENT'] = "PaymentCardPresent";
        TypeEnum[TypeEnum["BankAccount"] = 'BANK_ACCOUNT'] = "BankAccount";
        TypeEnum[TypeEnum["SwipedPaymentCard"] = 'SWIPED_PAYMENT_CARD'] = "SwipedPaymentCard";
    })(TypeEnum = PaymentInstrument.TypeEnum || (PaymentInstrument.TypeEnum = {}));
    let AccountTypeEnum;
    (function (AccountTypeEnum) {
        AccountTypeEnum[AccountTypeEnum["Checking"] = 'CHECKING'] = "Checking";
        AccountTypeEnum[AccountTypeEnum["Savings"] = 'SAVINGS'] = "Savings";
    })(AccountTypeEnum = PaymentInstrument.AccountTypeEnum || (PaymentInstrument.AccountTypeEnum = {}));
    let BankAccountValidationCheckEnum;
    (function (BankAccountValidationCheckEnum) {
        BankAccountValidationCheckEnum[BankAccountValidationCheckEnum["Valid"] = 'VALID'] = "Valid";
        BankAccountValidationCheckEnum[BankAccountValidationCheckEnum["Inconclusive"] = 'INCONCLUSIVE'] = "Inconclusive";
        BankAccountValidationCheckEnum[BankAccountValidationCheckEnum["NotAttempted"] = 'NOT_ATTEMPTED'] = "NotAttempted";
        BankAccountValidationCheckEnum[BankAccountValidationCheckEnum["Invalid"] = 'INVALID'] = "Invalid";
    })(BankAccountValidationCheckEnum = PaymentInstrument.BankAccountValidationCheckEnum || (PaymentInstrument.BankAccountValidationCheckEnum = {}));
})(PaymentInstrument = exports.PaymentInstrument || (exports.PaymentInstrument = {}));
