// @finix-payments/finix/lib/model/payoutProfileNet.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayoutProfileNet = void 0;
/**
* Details how `Payout Profiles` with type **NET** are configured.
*/
class PayoutProfileNet {
    static getAttributeTypeMap() {
        return PayoutProfileNet.attributeTypeMap;
    }
}
exports.PayoutProfileNet = PayoutProfileNet;
PayoutProfileNet.discriminator = undefined;
PayoutProfileNet.attributeTypeMap = [
    {
        "name": "frequency",
        "baseName": "frequency",
        "type": "PayoutProfileNet.FrequencyEnum"
    },
    {
        "name": "paymentInstrumentId",
        "baseName": "payment_instrument_id",
        "type": "string"
    },
    {
        "name": "rail",
        "baseName": "rail",
        "type": "PayoutProfileNet.RailEnum"
    },
    {
        "name": "submissionDelayDays",
        "baseName": "submission_delay_days",
        "type": "number"
    }
];
(function (PayoutProfileNet) {
    let FrequencyEnum;
    (function (FrequencyEnum) {
        FrequencyEnum[FrequencyEnum["Daily"] = 'DAILY'] = "Daily";
        FrequencyEnum[FrequencyEnum["Monthly"] = 'MONTHLY'] = "Monthly";
        FrequencyEnum[FrequencyEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FrequencyEnum = PayoutProfileNet.FrequencyEnum || (PayoutProfileNet.FrequencyEnum = {}));
    let RailEnum;
    (function (RailEnum) {
        RailEnum[RailEnum["NextDayAch"] = 'NEXT_DAY_ACH'] = "NextDayAch";
        RailEnum[RailEnum["SameDayAch"] = 'SAME_DAY_ACH'] = "SameDayAch";
    })(RailEnum = PayoutProfileNet.RailEnum || (PayoutProfileNet.RailEnum = {}));
})(PayoutProfileNet = exports.PayoutProfileNet || (exports.PayoutProfileNet = {}));
