// @finix-payments/finix/lib/model/payoutProfileGrossFees.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayoutProfileGrossFees = void 0;
/**
* Configures and details how fees get debited.
*/
class PayoutProfileGrossFees {
    static getAttributeTypeMap() {
        return PayoutProfileGrossFees.attributeTypeMap;
    }
}
exports.PayoutProfileGrossFees = PayoutProfileGrossFees;
PayoutProfileGrossFees.discriminator = undefined;
PayoutProfileGrossFees.attributeTypeMap = [
    {
        "name": "dayOfMonth",
        "baseName": "day_of_month",
        "type": "number"
    },
    {
        "name": "frequency",
        "baseName": "frequency",
        "type": "PayoutProfileGrossFees.FrequencyEnum"
    },
    {
        "name": "paymentInstrumentId",
        "baseName": "payment_instrument_id",
        "type": "string"
    },
    {
        "name": "rail",
        "baseName": "rail",
        "type": "PayoutProfileGrossFees.RailEnum"
    },
    {
        "name": "submissionDelayDays",
        "baseName": "submission_delay_days",
        "type": "number"
    }
];
(function (PayoutProfileGrossFees) {
    let FrequencyEnum;
    (function (FrequencyEnum) {
        FrequencyEnum[FrequencyEnum["Daily"] = 'DAILY'] = "Daily";
        FrequencyEnum[FrequencyEnum["Monthly"] = 'MONTHLY'] = "Monthly";
        FrequencyEnum[FrequencyEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FrequencyEnum = PayoutProfileGrossFees.FrequencyEnum || (PayoutProfileGrossFees.FrequencyEnum = {}));
    let RailEnum;
    (function (RailEnum) {
        RailEnum[RailEnum["NextDayAch"] = 'NEXT_DAY_ACH'] = "NextDayAch";
        RailEnum[RailEnum["SameDayAch"] = 'SAME_DAY_ACH'] = "SameDayAch";
    })(RailEnum = PayoutProfileGrossFees.RailEnum || (PayoutProfileGrossFees.RailEnum = {}));
})(PayoutProfileGrossFees = exports.PayoutProfileGrossFees || (exports.PayoutProfileGrossFees = {}));
