// @finix-payments/finix/lib/model/payoutProfileGrossPayouts.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayoutProfileGrossPayouts = void 0;
/**
* Configures and details how payouts get paid and credited.
*/
class PayoutProfileGrossPayouts {
    static getAttributeTypeMap() {
        return PayoutProfileGrossPayouts.attributeTypeMap;
    }
}
exports.PayoutProfileGrossPayouts = PayoutProfileGrossPayouts;
PayoutProfileGrossPayouts.discriminator = undefined;
PayoutProfileGrossPayouts.attributeTypeMap = [
    {
        "name": "frequency",
        "baseName": "frequency",
        "type": "PayoutProfileGrossPayouts.FrequencyEnum"
    },
    {
        "name": "paymentInstrumentId",
        "baseName": "payment_instrument_id",
        "type": "string"
    },
    {
        "name": "rail",
        "baseName": "rail",
        "type": "PayoutProfileGrossPayouts.RailEnum"
    },
    {
        "name": "submissionDelayDays",
        "baseName": "submission_delay_days",
        "type": "number"
    }
];
(function (PayoutProfileGrossPayouts) {
    let FrequencyEnum;
    (function (FrequencyEnum) {
        FrequencyEnum[FrequencyEnum["Daily"] = 'DAILY'] = "Daily";
        FrequencyEnum[FrequencyEnum["Monthly"] = 'MONTHLY'] = "Monthly";
        FrequencyEnum[FrequencyEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FrequencyEnum = PayoutProfileGrossPayouts.FrequencyEnum || (PayoutProfileGrossPayouts.FrequencyEnum = {}));
    let RailEnum;
    (function (RailEnum) {
        RailEnum[RailEnum["NextDayAch"] = 'NEXT_DAY_ACH'] = "NextDayAch";
        RailEnum[RailEnum["SameDayAch"] = 'SAME_DAY_ACH'] = "SameDayAch";
    })(RailEnum = PayoutProfileGrossPayouts.RailEnum || (PayoutProfileGrossPayouts.RailEnum = {}));
})(PayoutProfileGrossPayouts = exports.PayoutProfileGrossPayouts || (exports.PayoutProfileGrossPayouts = {}));
