// @finix-payments/finix/lib/model/settlement.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settlement = void 0;
/**
*
*/
class Settlement {
    static getAttributeTypeMap() {
        return Settlement.attributeTypeMap;
    }
}
exports.Settlement = Settlement;
Settlement.discriminator = undefined;
Settlement.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "application",
        "baseName": "application",
        "type": "string"
    },
    {
        "name": "currency",
        "baseName": "currency",
        "type": "Currency"
    },
    {
        "name": "destination",
        "baseName": "destination",
        "type": "string"
    },
    {
        "name": "fundsFlow",
        "baseName": "funds_flow",
        "type": "string"
    },
    {
        "name": "identity",
        "baseName": "identity",
        "type": "string"
    },
    {
        "name": "merchantId",
        "baseName": "merchant_id",
        "type": "string"
    },
    {
        "name": "netAmount",
        "baseName": "net_amount",
        "type": "number"
    },
    {
        "name": "paymentType",
        "baseName": "payment_type",
        "type": "string"
    },
    {
        "name": "processor",
        "baseName": "processor",
        "type": "string"
    },
    {
        "name": "status",
        "baseName": "status",
        "type": "Settlement.StatusEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "totalAmount",
        "baseName": "total_amount",
        "type": "number"
    },
    {
        "name": "totalFee",
        "baseName": "total_fee",
        "type": "number"
    },
    {
        "name": "totalFees",
        "baseName": "total_fees",
        "type": "number"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "Settlement.TypeEnum"
    },
    {
        "name": "links",
        "baseName": "_links",
        "type": "SettlementLinks"
    }
];
(function (Settlement) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["Approved"] = 'APPROVED'] = "Approved";
        StatusEnum[StatusEnum["AwaitingApproval"] = 'AWAITING_APPROVAL'] = "AwaitingApproval";
        StatusEnum[StatusEnum["Pending"] = 'PENDING'] = "Pending";
    })(StatusEnum = Settlement.StatusEnum || (Settlement.StatusEnum = {}));
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["MerchantRevenue"] = 'MERCHANT_REVENUE'] = "MerchantRevenue";
        TypeEnum[TypeEnum["PlatformFee"] = 'PLATFORM_FEE'] = "PlatformFee";
        TypeEnum[TypeEnum["PartnerFee"] = 'PARTNER_FEE'] = "PartnerFee";
        TypeEnum[TypeEnum["Noop"] = 'NOOP'] = "Noop";
        TypeEnum[TypeEnum["Merchant"] = 'MERCHANT'] = "Merchant";
        TypeEnum[TypeEnum["Application"] = 'APPLICATION'] = "Application";
        TypeEnum[TypeEnum["Platform"] = 'PLATFORM'] = "Platform";
    })(TypeEnum = Settlement.TypeEnum || (Settlement.TypeEnum = {}));
})(Settlement = exports.Settlement || (exports.Settlement = {}));
