// @finix-payments/finix/lib/model/subTypeTransfer.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubTypeTransfer = void 0;
var SubTypeTransfer;
(function (SubTypeTransfer) {
    SubTypeTransfer[SubTypeTransfer["Api"] = 'API'] = "Api";
    SubTypeTransfer[SubTypeTransfer["ApplicationFee"] = 'APPLICATION_FEE'] = "ApplicationFee";
    SubTypeTransfer[SubTypeTransfer["Dispute"] = 'DISPUTE'] = "Dispute";
    SubTypeTransfer[SubTypeTransfer["MerchantCredit"] = 'MERCHANT_CREDIT'] = "MerchantCredit";
    SubTypeTransfer[SubTypeTransfer["MerchantCreditAdjustment"] = 'MERCHANT_CREDIT_ADJUSTMENT'] = "MerchantCreditAdjustment";
    SubTypeTransfer[SubTypeTransfer["MerchantDebit"] = 'MERCHANT_DEBIT'] = "MerchantDebit";
    SubTypeTransfer[SubTypeTransfer["MerchantDebitAdjustment"] = 'MERCHANT_DEBIT_ADJUSTMENT'] = "MerchantDebitAdjustment";
    SubTypeTransfer[SubTypeTransfer["PlatformCredit"] = 'PLATFORM_CREDIT'] = "PlatformCredit";
    SubTypeTransfer[SubTypeTransfer["PlatformCreditAdjustment"] = 'PLATFORM_CREDIT_ADJUSTMENT'] = "PlatformCreditAdjustment";
    SubTypeTransfer[SubTypeTransfer["PlatformDebit"] = 'PLATFORM_DEBIT'] = "PlatformDebit";
    SubTypeTransfer[SubTypeTransfer["PlatformDebitAdjustment"] = 'PLATFORM_DEBIT_ADJUSTMENT'] = "PlatformDebitAdjustment";
    SubTypeTransfer[SubTypeTransfer["PlatformFee"] = 'PLATFORM_FEE'] = "PlatformFee";
    SubTypeTransfer[SubTypeTransfer["SettlementMerchant"] = 'SETTLEMENT_MERCHANT'] = "SettlementMerchant";
    SubTypeTransfer[SubTypeTransfer["SettlementNoop"] = 'SETTLEMENT_NOOP'] = "SettlementNoop";
    SubTypeTransfer[SubTypeTransfer["SettlementPartner"] = 'SETTLEMENT_PARTNER'] = "SettlementPartner";
    SubTypeTransfer[SubTypeTransfer["SettlementPlatform"] = 'SETTLEMENT_PLATFORM'] = "SettlementPlatform";
    SubTypeTransfer[SubTypeTransfer["SplitPayout"] = 'SPLIT_PAYOUT'] = "SplitPayout";
    SubTypeTransfer[SubTypeTransfer["SplitPayoutAdjustment"] = 'SPLIT_PAYOUT_ADJUSTMENT'] = "SplitPayoutAdjustment";
})(SubTypeTransfer = exports.SubTypeTransfer || (exports.SubTypeTransfer = {}));
