// @finix-payments/finix/lib/model/updateDeviceRequest.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateDeviceRequest = void 0;
class UpdateDeviceRequest {
    static getAttributeTypeMap() {
        return UpdateDeviceRequest.attributeTypeMap;
    }
}
exports.UpdateDeviceRequest = UpdateDeviceRequest;
UpdateDeviceRequest.discriminator = undefined;
UpdateDeviceRequest.attributeTypeMap = [
    {
        "name": "action",
        "baseName": "action",
        "type": "UpdateDeviceRequest.ActionEnum"
    },
    {
        "name": "activationCode",
        "baseName": "activation_code",
        "type": "string"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "idleMessage",
        "baseName": "idle_message",
        "type": "string"
    }
];
(function (UpdateDeviceRequest) {
    let ActionEnum;
    (function (ActionEnum) {
        ActionEnum[ActionEnum["Reboot"] = 'REBOOT'] = "Reboot";
        ActionEnum[ActionEnum["Deactivate"] = 'DEACTIVATE'] = "Deactivate";
        ActionEnum[ActionEnum["Activate"] = 'ACTIVATE'] = "Activate";
        ActionEnum[ActionEnum["CreateIdleMessage"] = 'CREATE_IDLE_MESSAGE'] = "CreateIdleMessage";
    })(ActionEnum = UpdateDeviceRequest.ActionEnum || (UpdateDeviceRequest.ActionEnum = {}));
})(UpdateDeviceRequest = exports.UpdateDeviceRequest || (exports.UpdateDeviceRequest = {}));
