// @finix-payments/finix/lib/model/updatePayoutProfileGrossFees.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdatePayoutProfileGrossFees = void 0;
/**
* Configures the details of how fees get debited.
*/
class UpdatePayoutProfileGrossFees {
    static getAttributeTypeMap() {
        return UpdatePayoutProfileGrossFees.attributeTypeMap;
    }
}
exports.UpdatePayoutProfileGrossFees = UpdatePayoutProfileGrossFees;
UpdatePayoutProfileGrossFees.discriminator = undefined;
UpdatePayoutProfileGrossFees.attributeTypeMap = [
    {
        "name": "dayOfMonth",
        "baseName": "day_of_month",
        "type": "number"
    },
    {
        "name": "frequency",
        "baseName": "frequency",
        "type": "UpdatePayoutProfileGrossFees.FrequencyEnum"
    },
    {
        "name": "paymentInstrumentId",
        "baseName": "payment_instrument_id",
        "type": "string"
    },
    {
        "name": "rail",
        "baseName": "rail",
        "type": "UpdatePayoutProfileGrossFees.RailEnum"
    },
    {
        "name": "submissionDelayDays",
        "baseName": "submission_delay_days",
        "type": "number"
    }
];
(function (UpdatePayoutProfileGrossFees) {
    let FrequencyEnum;
    (function (FrequencyEnum) {
        FrequencyEnum[FrequencyEnum["Daily"] = 'DAILY'] = "Daily";
        FrequencyEnum[FrequencyEnum["Monthly"] = 'MONTHLY'] = "Monthly";
        FrequencyEnum[FrequencyEnum["Custom"] = 'CUSTOM'] = "Custom";
    })(FrequencyEnum = UpdatePayoutProfileGrossFees.FrequencyEnum || (UpdatePayoutProfileGrossFees.FrequencyEnum = {}));
    let RailEnum;
    (function (RailEnum) {
        RailEnum[RailEnum["NextDayAch"] = 'NEXT_DAY_ACH'] = "NextDayAch";
        RailEnum[RailEnum["SameDayAch"] = 'SAME_DAY_ACH'] = "SameDayAch";
    })(RailEnum = UpdatePayoutProfileGrossFees.RailEnum || (UpdatePayoutProfileGrossFees.RailEnum = {}));
})(UpdatePayoutProfileGrossFees = exports.UpdatePayoutProfileGrossFees || (exports.UpdatePayoutProfileGrossFees = {}));
