// @finix-payments/finix/lib/model/user.js
"use strict";
/**
 * Finix API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
class User {
    static getAttributeTypeMap() {
        return User.attributeTypeMap;
    }
}
exports.User = User;
User.discriminator = undefined;
User.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "string"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "updatedAt",
        "baseName": "updated_at",
        "type": "Date"
    },
    {
        "name": "enabled",
        "baseName": "enabled",
        "type": "boolean"
    },
    {
        "name": "identity",
        "baseName": "identity",
        "type": "string"
    },
    {
        "name": "password",
        "baseName": "password",
        "type": "string"
    },
    {
        "name": "role",
        "baseName": "role",
        "type": "User.RoleEnum"
    },
    {
        "name": "tags",
        "baseName": "tags",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "links",
        "baseName": "_links",
        "type": "UserLinks"
    }
];
(function (User) {
    let RoleEnum;
    (function (RoleEnum) {
        RoleEnum[RoleEnum["Admin"] = 'ROLE_ADMIN'] = "Admin";
        RoleEnum[RoleEnum["Platform"] = 'ROLE_PLATFORM'] = "Platform";
        RoleEnum[RoleEnum["Partner"] = 'ROLE_PARTNER'] = "Partner";
        RoleEnum[RoleEnum["Merchant"] = 'ROLE_MERCHANT'] = "Merchant";
    })(RoleEnum = User.RoleEnum || (User.RoleEnum = {}));
})(User = exports.User || (exports.User = {}));
